### PR TITLE
dag: fix ReverseDepthFirstWalk when nodes remove themselves

### LINKED ISSUE #7378: 0.7.0 regression: "unknown variable accessed" in destroy
### Terraform Version

I can reproduce this with today's HEAD (ebb9b7a) and with 0.7.0-rc2. **The issue does not occur with 0.6.16**: this is a 0.7.0 regression.

### Affected Resource(s)

Seems like core.

### Terraform Configuration Files

You can get this by `git clone https://github.com/glasser/terraform-destroy-bug` and here is the code inline:

```hcl
# top.tf
module "middle" {
  source = "./middle"
  param  = "foo"
}
```

```hcl
# middle/middle.tf
variable param {}

resource "null_resource" "n" {}

module "bottom" {
  source       = "./bottom"
  bottom_param = "${var.param}"
}
```

```hcl
# middle/bottom/bottom.tf
variable bottom_param {}
```

### Steps to Reproduce
1. `terraform get`
2. `terraform apply`
3. `terraform destroy -force`

### Expected Behavior

This should create and destroy the single `null_resource`.

### Actual Behavior

The `destroy -force` fails with:

```
module.middle.null_resource.n: Refreshing state... (ID: 8515402792609950867)
null_resource.n: Destroying...
null_resource.n: Destruction complete
Error applying plan:

1 error(s) occurred:

* 1:3: unknown variable accessed: var.param in:

${var.param}

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

It does not matter which of the three modules contains the `null_resource` (and the issue is not specific to `null_resource`); it just appears that the configuration must contain at least one resource somewhere for the bug to be triggered.

### Debug Output

`TF_LOG=DEBUG`: https://gist.github.com/glasser/9ad8b72709072ceddae2ae44111c9a94
`TF_LOG=TRACE`: https://gist.github.com/glasser/e830d64ec640c80b63c0614cc4123406

### References

This seems similar to #7131 (fixed by @phinze and reviewed by @jen20); however, the fix there does not fix this situation.  (I think we were also encountering whatever #7131 fixed in our stack, as upgrading to get that fix did reduce the number of errors of this form that we saw, but not entirely.)

### PR DESCRIPTION
**Pull request by phinze**
The report in #7378 led us into a deep rabbit hole that turned out to
expose a bug in the graph walk implementation being used by the
`NoopTransformer`. The problem ended up being when two nodes in a single
dependency chain both reported `Noop() -> true` and needed to be
removed. This was breaking the walk and preventing the second node from
ever being visited.

Fixes #7378

### CODE REVIEW

**Comment by phinze** [comment]
cc @jen20 / @jbardin for 👀

File: dag/dag.go [hunk 1]
```diff
@@ -346,6 +341,11 @@ func (g *AcyclicGraph) ReverseDepthFirstWalk(start []Vertex, f DepthWalkFunc) er
 				Depth:  current.Depth + 1,
 			})
 		}
+
+		// Visit the current node
+		if err := f(current.Vertex, current.Depth); err != nil {
+			return err
+		}
```
Comment (phinze) [inline dag/dag.go]: The other graph walk has this same problem, but I wanted to address it in a separate PR w/ its own tests.

Comment (phinze) [inline dag/dag.go]: Also are these depth first? I'm suspicious. Will have to double check this and potentially rename.

**Comment by phinze** [comment]
I'm going to merge this so I can start the RC3 build. Just manually verified that this fixes the attached issue. Still would like a review from both of you @jen20 and @jbardin when you get the chance.

**Comment by glasser** [comment]
Woo! I'm going to try this on my real case ASAP to see if it actually fixes the issue in our full setup.

**Comment by glasser** [comment]
Great, this fixes not only my minimized reproduction but my real use case as well!

File: terraform/test-fixtures/apply-destroy-nested-module-with-attrs/middle/bottom/bottom.tf [hunk 2]
```diff
@@ -1 +1,5 @@
-variable "bottom_param" {}
+variable bottom_param {}
+
+resource "null_resource" "bottom" {
+  value = "${var.bottom_param}"
```
Comment (glasser) [inline terraform/test-fixtures/apply-destroy-nested-module-with-attrs/middle/bottom/bottom.tf]: Just a question to help me form good reproductions in the future: What's this `value` on `null_resource`? Can you just put whatever keys you want on `null_resource`, or is `value` something special?

Comment (phinze) [inline terraform/test-fixtures/apply-destroy-nested-module-with-attrs/middle/bottom/bottom.tf]: Good question! In this test context, there is not an actual provider present. We [wire in a fake provider](https://github.com/hashicorp/terraform/blob/master/terraform/context_apply_test.go#L4816) and give it whatever name we want.

If you're actually asking about the proper `null_resource` behavior, it only supports `triggers` which is an arbitrary map of values that when changed will cause its provisioners to be re-run.

Hope his helps to clarify!

**Pull request merged**